File: springboot-02-ioccontainer/src/main/java/com/linkedbear/springboot/assemble/d_spi/JdkSpiApplication.java
package com.linkedbear.springboot.assemble.d_spi;

import com.linkedbear.springboot.assemble.d_spi.bean.DemoDao;
import java.util.ServiceLoader;

public class JdkSpiApplication {
    
    public static void main(String[] args) {
        ServiceLoader<DemoDao> serviceLoader = ServiceLoader.load(DemoDao.class);
        serviceLoader.iterator().forEachRemaining(System.out::println);
    }
}

File: springboot-02-ioccontainer/src/main/kotlin/com/linkedbear/springboot/assemble/d_spi/JdkSpiApplication.kt
package com.linkedbear.springboot.assemble.d_spi

import com.linkedbear.springboot.assemble.d_spi.bean.DemoDao
import java.util.ServiceLoader

class JdkSpiApplication {
    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val serviceLoader = ServiceLoader.load(DemoDao::class.java)
            serviceLoader.forEach {
                println(it)
            }
        }
    }
}